packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -0,0 +1,280 @@
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//
+// ************************************************************************
+// @HEADER
+
+#ifndef TPETRA_DETAILS_RESIDUAL_HPP
+#define TPETRA_DETAILS_RESIDUAL_HPP
+
+#include "TpetraCore_config.h"
+#include "Tpetra_CrsMatrix.hpp"
+#include "Tpetra_MultiVector.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Teuchos_ScalarTraits.hpp"
+#include "Tpetra_Details_Behavior.hpp"
+#include "Tpetra_Details_Profiling.hpp"
+
+/// \file Tpetra_Details_residual.hpp
+/// \brief Functions that allow for fused residual calculation.
+/// \warning This file, and its contents, are implementation details
+///   of Tpetra.  The file itself or its contents may disappear or
+///   change at any time.-
+
+namespace Tpetra {
+  namespace Details {
+
+template<class SC, class LO, class GO, class NO>
+void localResidual(const CrsMatrix<SC,LO,GO,NO> &  A,
+                   const MultiVector<SC,LO,GO,NO> & X,
+                   const MultiVector<SC,LO,GO,NO> & B,
+                   MultiVector<SC,LO,GO,NO> & R) {
+  using Tpetra::Details::ProfilingRegion;
+  using Teuchos::NO_TRANS;
+  ProfilingRegion regionLocalApply ("Tpetra::CrsMatrix::localResidual");
+
+  auto X_lcl = X.getLocalViewDevice ();
+  auto B_lcl = B.getLocalViewDevice ();
+  auto R_lcl = R.getLocalViewDevice ();
+  auto lclMatrix_ = A.getLocalMatrix ();
+
+  const bool debug = ::Tpetra::Details::Behavior::debug ();
+  if (debug) {
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (X.getNumVectors () != R.getNumVectors (), std::runtime_error,
+       "X.getNumVectors() = " << X.getNumVectors () << " != "
+       "R.getNumVectors() = " << R.getNumVectors () << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (X.getNumVectors () != R.getNumVectors (), std::runtime_error,
+       "X.getNumVectors() = " << X.getNumVectors () << " != "
+       "R.getNumVectors() = " << R.getNumVectors () << ".");
+
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (X.getLocalLength () !=
+       A.getColMap ()->getNodeNumElements (), std::runtime_error,
+       "X has the wrong number of local rows.  "
+       "X.getLocalLength() = " << X.getLocalLength () << " != "
+       "A.getColMap()->getNodeNumElements() = " <<
+       A.getColMap ()->getNodeNumElements () << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (R.getLocalLength () !=
+       A.getRowMap ()->getNodeNumElements (), std::runtime_error,
+       "R has the wrong number of local rows.  "
+       "R.getLocalLength() = " << R.getLocalLength () << " != "
+       "A.getRangeMap()->getNodeNumElements() = " <<
```

Comment:
This is post-Import, pre-Export, so we should be checking getRowMap here.   Note that this was copied from apply, so apply  might be wrong here.

Reply:
Fixed.  Apply looks clean.